### PR TITLE
Optimize Dex#getEffect by caching the result.

### PR DESCRIPTION
**Pull request by scheibo**
I've been taking a look at optimizing the simulator (see [pkmn.cc/optimize](https://pkmn.cc/optimize) - comments welcome), and `getEffect` seems ripe for improvements. I can't confirm this caching is 100% safe - I don't know if `Effect`s are supposed to be immutable or not, but the 'tests pass', for whatever that's worth. There's already caching for `Move`/`Ability`/`Item` etc, but `getEffect` only very rarely hits those code paths, so it's actually missing out on most of the benefit. 

This change should result in a **~30% increase in overall battle performance** over repeated runs, but if this caching would break things theres still some other options for restructuring the method for performance, as outlined in the doc. I think there's more performance gains to be had by refactoring `getRelevantEffects`, but it would require a substantially larger change architecturally, so I'd hope to get some buy in before tackling that.

### CODE REVIEW

**Comment by Zarel** [comment]
Wow! That's an impressive amount of research into performance optimization.

**Comment by Zarel** [comment]
`getRelevantEffects` is unsurprising as a culprit.

In the past, when I've profiled, I've seen a lot of time spent in `toId`. Most `toId` calls can be optimized out because we mostly pass around IDs these days, but for safety reasons we'd need to actually migrate to Real TypeScript so we can use the `ID` type like we do in the client, to make sure nothing breaks.

File: sim/dex.js [hunk 1]
```diff
@@ -458,10 +460,8 @@ class ModdedDex {
 		moveCopy.hit = 0;
 		return moveCopy;
 	}
+
 	/**
-	 * While this function can technically return any kind of effect at
-	 * all, that's not a feature TypeScript needs to know about.
-	 *
```
Comment (Zarel) [inline sim/dex.js]: Why is this removed? It's still correct.

File: sim/dex.js [hunk 2]
```diff
@@ -473,6 +473,22 @@ class ModdedDex {
...
+		return cached;
+	}
+
+	/**
+	* While this function can technically return any kind of effect at
+	* all, that's not a feature TypeScript needs to know about.
+	* @param {string} name
+	* @return {PureEffect}
+	*/
+	getEffectInner(name) {
```
Comment (Zarel) [inline sim/dex.js]: I'd normally not split this into two functions (something something call stack height). Why aren't you doing the same caching methods the other `get` functions do?

File: sim/dex.js (hunk 1 above)
Comment (Zarel) [reply sim/dex.js]: Oh, I see, it's moved to the inner function. If you must have an inner function, I think the comment still belongs here.

Comment (scheibo) [reply sim/dex.js]: I left this on `getEffectInner`, but didn't copy the comment to `getEffect`. Intuitively it makes sense on `getEffect` because thats the 'public' method people should call, but at the same time, its not the method using '`// @ts-ignore`' all over the place, so I thought it made more sense to leave it with `getEffectInner`. I'm happy to place it wherever you think appropriate.

(I'm curious as to why `getEffect` doesn't just return `Effect` instead of `PureEffect`?)

File: sim/dex.js (hunk 2 above)
Comment (scheibo) [reply sim/dex.js]: Mainly due to the ugliness of the return type here and all the `// @ts-ignore`. With `effect` at the bottom its pretty easy, but getMove, getAbility etc return Effects, not PureEffects so I need to do more ugly ignoring and it just seemed easier to contain breaking the types into one method. I think I was also influenced by looking at optimizing getRelevantEffects/getRelevantEffectsInner, but thats a different pattern (one adds additional parameters) and in a different class, so I'll follow the more standard pattern here.

On a side note: interesting you're worried about call stack height as opposed to aiming for small, easily inlineable methods (in the past, JS engines would even count comments and whitespace against you WRT inlining, so short methods were king. I'm not sure to what extent that still holds, but I believe its still relevant). Presumably you've been bitten by stack height in the past, judging by the bail out logic in `runEvent` and the super long methods.

**Comment by Zarel** [comment]
I'm surprised getting the format is a big problem, though; I'm pretty sure formats are currently cached. Once when we accidentally stopped caching formats, we had _huge_ lag issues, which is why I'm surprised you're telling me there's still this many calls.

File: sim/dex.js (hunk 1 above)
Comment (Zarel) [reply sim/dex.js]: "That's not a feature TypeScript needs to know about" is the hint here. It returns `PureEffect` because most calls to `getEffect` expect (and are guaranteed) a `PureEffect`, so returning `PureEffect` avoids a lot of casting on the caller side.

**Comment by scheibo** [comment]
I didn't break down how long each respective call type takes (ie. while there certainly are a lot of calls to format, I didn't determine how fast each call is - they could very well be cached). At the same time, yes, the *format* may be cached, but not the *effect* for the format - you still `new` up an object every time, and that isn't free (allocation and GC).

**Comment by Zarel** [comment]
By the way, this is pretty good timing! I only recently started enforcing immutable PureEffects using TypeScript, so this is much less dangerous than it would have been before that. I think it's mergeable as-is.

**Comment by Zarel** [comment]
Where's the format's effect used, anyway? In theory it's not used at all, in a random battle.

File: sim/dex.js (hunk 2 above)
Comment (Zarel) [reply sim/dex.js]: Yeah, I've had some trouble with stack traces not containing enough information to diagnose bugs in the past. For performance reasons, stack traces are capped at a specific height. That, and I have a general rule of thumb that if you can't think of a better function name than `[...]Inner`, it shouldn't be its own function.

File: sim/dex.js (hunk 1 above)
Comment (Zarel) [reply sim/dex.js]: (And because we're using janky TS-in-JS instead of real TypeScript, casting is really hard.)

**Comment by scheibo** [comment]
> Where's the format's effect used, anyway? In theory it's not used at all, in a random battle.

I'm not sure that the format's effects are used - as I understand it, `getRelevantsEffects` just tries to find effects of a given type, independent of whether they are actually going to get used, no? I think the reason we see it popping up so frequently, even in random battle, is that format is where things like 'Sleep Clause Mod' are implemented, which comes up pretty much in all battle types.

> In the past, when I've profiled, I've seen a lot of time spent in `toId`. Most `toId` calls can be optimized out because we mostly pass around IDs these days, but for safety reasons we'd need to actually migrate to Real TypeScript so we can use the `ID` type like we do in the client, to make sure nothing breaks.

Is there a roadmap/design doc for migrating to full Typescript. It's obviously a large change, but I'm wondering if it's something you've started working on (having typed most of the code with JSDoc surely indicates as such)? I'd be interested in helping out.

> `getRelevantEffects` is unsurprising as a culprit.

Do you have ideas for how to optimize it? What do you think about the proposal sketched in the doc for doing work up front when effects are added to/removed from objects as opposed to querying the effects (potentially multiple times over their lifetime)? There's some nuances involved, and I would have to benchmark the real code once implemented, but I think its worth pursuing. If it sounds like something you'd consider merging I'll dive deeper into it, but if it would be too disruptive architecturally I'll try to think of other ideas.

**Comment by Zarel** [comment]
Your proposal sounds like a good direction.

There's no work started on TypeScript migration yet, but the plan is pretty straightforward: just rename the files to `.ts` one at a time, and add a compiler to `./pokemon-showdown`.
